IPlug/AUv3/iOSApp/IPlugIOSMIDI.mm

```diff
@@ -0,0 +1,230 @@
+
+// Static Helpers
+
+//static
+ItemCount IPlugIOSMIDI::GetNumSourcesOrDestinations(ERoute route)
+{
+  if (route == iplug::kInput)
+    return MIDIGetNumberOfSources();
+  else
+    return MIDIGetNumberOfDestinations();
+}
+
+//static
+MIDIEndpointRef IPlugIOSMIDI::GetEndpoint(ItemCount idx, ERoute route)
+{
+  if (route == iplug::kInput)
+    return MIDIGetSource(idx);
+  else
+    return MIDIGetDestination(idx);
+}
+
+//static
+CFStringRef IPlugIOSMIDI::CreateNameFromMIDEndpoint(MIDIEndpointRef endpoint)
+{
+  CFStringRef names[2];
+  MIDIEntityRef entity;
+  MIDIDeviceRef device;
+  
+  MIDIEndpointGetEntity(endpoint, &entity);
+  MIDIEntityGetDevice(entity, &device);
+  
+  MIDIObjectGetStringProperty(device, kMIDIPropertyName, names + 0);
+  MIDIObjectGetStringProperty(endpoint, kMIDIPropertyName, names + 1);
+  
+  CFStringRef name;
+  
+  if (CFStringCompare(names[0], names[1], kCFCompareCaseInsensitive) == kCFCompareEqualTo)
+  {
+    name = CFStringCreateCopy(NULL, names[0]);
+  }
+  else
+  {
+    CFStringRef separator = CFStringCreateWithCString(NULL, ": ", kCFStringEncodingUTF8);
+    CFArrayRef array = CFArrayCreate(NULL, (const void **) names, 2, NULL);
+    
+    name = CFStringCreateByCombiningStrings(NULL, array, separator);
+    
+    CFRelease(array);
+    CFRelease(separator);
+  }
+  
+  return name;
+}
+
+//static
+CFStringRef IPlugIOSMIDI::CreateNameFromIndex(ItemCount idx, ERoute route)
+{
+  return CreateNameFromMIDEndpoint(GetEndpoint(idx, route));
+}
+
+//static
+void IPlugIOSMIDI::GetNameFromIndex(WDL_String &string, int idx, ERoute route)
+{
+  char cString[2048];
+  
+  CFStringRef str = CreateNameFromIndex(idx, route);
+  CFStringGetCString(str, cString, sizeof(cString), kCFStringEncodingUTF8);
+  CFRelease(str);
+  string.Set(cString);
+}
+
+//static
+long IPlugIOSMIDI::GetIndexFromName(CFStringRef name, ERoute route)
+{
+  auto numEndPoints = GetNumSourcesOrDestinations(route);
+  long idx = -1;
+  
+  for (long i = 0; i < numEndPoints && idx < 0; i++)
+  {
+    auto str = CreateNameFromMIDEndpoint(GetEndpoint(i, route));
+    if (CFStringCompare(name, str, 0) == kCFCompareEqualTo)
+      idx = i;
+    CFRelease(str);
+  }
+  
+  return idx;
+}
+
+//static
+long IPlugIOSMIDI::GetIndexFromName(WDL_String name, ERoute route)
+{
+  CFStringRef str = CFStringCreateWithCString(NULL, name.Get(), kCFStringEncodingUTF8);
+  long idx = GetIndexFromName(str, route);
+  CFRelease(str);
+  
+  return idx;
+}
+
+//static
+void IPlugIOSMIDI::SetMidiPort(const char *name, ERoute route)
+{
+  NSDictionary* dic = @{@"name": @(name), @"direction": @(route == iplug::kInput ? "source" : "destination")};
+  [[NSNotificationCenter defaultCenter] postNotificationName:@"SetMIDIPort" object:nil userInfo:dic];
+}
+
+//static
+void IPlugIOSMIDI::GetMidiPort(WDL_String& name, ERoute route)
+{
+  NSDictionary* dic = @{@"name": [NSValue valueWithPointer:&name], @"direction": @(route == iplug::kInput ? "source" : "destination")};
```

Comment:
Storing a pointer to the WDL_String here and setting the value in the notification handler feels wrong here... trying to think of other options @AlexHarker

Reply:
Yes - I can see why that's not ideal, but we are picking it up at the other end and the idea was to be able to get the value out here - I wouldn't want to pass a pointer to an plugin-developer in this way, but here the framework is using the notification system to pass around data. As far as I am aware (and I am pretty sure I would have tested before using this approach) the notification system is immediate and synchronous, so it should be safe - it's just that the notification system is the best way to communicate between app and plugin (given that the framework has no notion of the wrapping app). This was based on the BT Midi notification stuff - if there's a better way or a way that involves putting a hook in the plugin base etc. then I'd be ok with that.